### PR TITLE
Wallet API

### PR DESCRIPTION
**Pull request by pratikloya**
## Motivation

Alternative to Blockchain-API service for providing graphql query response.

## Changelog
* Added new elixir app - walletapi in app folder. 
* Modified Docker file to expose 4002 port.
* Updated Make file for local development.
* Added test cases covering the correctness of the functions and also the response of graphql queries.

### CODE REVIEW

File: apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        DateTime.utc_now()
+      else
+        DateTime.from_unix!(timestamp, :millisecond)
+      end
+
+    # is08601 format : YYYY-MM-DD
+    date = Date.to_iso8601(time_in_ms)
+
+    # if given time within last 24 hours - get current exchange rates from API.
+    if DateTime.diff(DateTime.utc_now(), time_in_ms) < 24 * 60 * 60 do
```
Comment (carterqw2) [inline apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex:44]: Hmm, does it mean that within the current 24 hours time frame every request will result in an API request?

File: apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  end
+
+  defp get_data_from_api(source_currency_code, date) do
+    path = "/historical?"
+    access_key = System.get_env("CURRENCY_LAYER_ACCESS_KEY")
+    req_body = URI.encode_query(access_key: access_key, source: source_currency_code, date: date)
+
+    case HTTPoison.get(@api_url <> path <> req_body) do
+      {:ok, %HTTPoison.Response{body: body, status_code: 200}} ->
+        response = Poison.decode!(body)
```
Comment (carterqw2) [inline apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex:87]: Just curious how it works in Elixir world, if you call a bang-function, are you supposed to handle the raised exception? And if you don't handle it here, are you supposed to add an exclamation mark to your own function or not?

File: apps/walletapi/lib/walletapi/schema/types.ex [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    field(:node, :token_transaction)
+    field(:cursor, non_null(:string))
+  end
+
+  defp resolve_transaction_type(obj, _) do
+    if(obj.type == :exchange) do
+      :token_exchange
+    else
+      if(
+        obj.type == :received ||
```
Comment (carterqw2) [inline apps/walletapi/lib/walletapi/schema/types.ex:162]: Can we use `Enum.member?` here perhaps?

File: apps/walletapi/config/test.exs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+# to provide built-in test partitioning in CI environment.
+# Run `mix help test` for more information.
+# config :walletapi, Explorer.Repo,
+#   pool: Ecto.Adapters.SQL.Sandbox
+
+# We don't run a server during test. If one is required,
+# you can enable the server option below.
+config :walletapi, WalletApi.Endpoint,
+  http: [port: System.get_env("WALLETAPI_PORT")],
+  secret_key_base: "27Swe6KtEtmN37WyEYRjKWyxYULNtrxlkCEKur4qoV+Lwtk8lafsR16ifz1XBBYj",
```
[the remark this answers is not in the document]
Comment (pratikloya) [reply apps/walletapi/config/test.exs:15]: Its the same key used in block_scout_web/config/test.ex and is already public.

File: apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex (hunk 1 above)
Comment (pratikloya) [reply apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex:44]: Yes, took the reference from blockchain-api. We are fetching the recent exchange_rates for transaction that happened within the last 24hrs.

File: apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex (hunk 2 above)
Comment (pratikloya) [reply apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex:87]: Even if we use bang_function we still need to catch the exception. And the problem with the currency_layer api is that it returns a 200 status even for incorrect request hence additional checks are performed in the return_response() function to validate the response.

File: apps/walletapi/lib/walletapi/schema/types.ex (hunk 3 above)
Comment (pratikloya) [reply apps/walletapi/lib/walletapi/schema/types.ex:162]: I may be wrong but I guess Enum.member? function will check if the value is present or not if its a list but the type value of obj is an atom.

File: apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex (hunk 1 above)
Comment (carterqw2) [reply apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex:44]: But if I'm reading your code correctly, if the time is within 24 hours, you'll make an API call every time to fetch the rate, unlike blockchain API code that would make a call once and cache it for 12 hours (https://github.com/celo-org/celo-monorepo/blob/master/packages/blockchain-api/src/currencyConversion/ExchangeRateAPI.ts#L47). Am I missing something? Will ConCache do it behind the scenes as well? It's a bit implicit, so hard to say what would be the behavior in this case.

File: apps/walletapi/config/test.exs (hunk 4 above)
[the remark this answers is not in the document]
Comment (carterqw2) [reply apps/walletapi/config/test.exs:15]: Gotcha

File: apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      get_data_from_cache(source_currency_code, date)
+    end
+  end
+
+  defp get_data_from_cache(source_currency_code, date) do
+    # global_ttl set to 1 year to store historical exchange rates
+    # *****************TODO******************
+    # Concache is in-memory cache and gets reset if the pod is cleaned
+    # Add another cache layer (GoogleCloud Cache) to store the values
+    # If in-memory cache is cleaned, get the rates from GoogleClousCache
```
[the remark this answers is not in the document]
Comment (pratikloya) [reply apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex]: Thanks, corrected the error

File: apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex (hunk 1 above)
Comment (pratikloya) [reply apps/walletapi/lib/walletapi/resolver/currency_conversion/exchange_rate_api.ex]: Modified the code to save the exchange rates for 12 hours for queries with time less than 24hrs.

File: apps/walletapi/lib/walletapi/schema/types.ex (hunk 3 above)
Comment (pratikloya) [reply apps/walletapi/lib/walletapi/schema/types.ex]: Modified the code. Now using Enum.member? to check the value from a list.

**Pull request closed without merging**